<h1>Pre-procesamiento y limpieza de Datos</h1>

In [1]:
'''Importar la libreria para la realizacion de este laboratorio.'''
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 

In [2]:
'''Obtener el Dataset para mostrar todo su contenido.'''
df = pd.read_csv("data_act_01.csv",sep=";")
df.sample(10)

CrimeId OriginalCrimeTypeName          OffenseDate CallTime  \
8289  160952782          Traffic Stop  2016-04-04T00:00:00    16:47   
404   160920536    Suspicious Vehicle  2016-04-01T00:00:00     6:58   
7000  160950420    Homeless Complaint  2016-04-04T00:00:00     5:58   
3617  160932304    Homeless Complaint  2016-04-02T00:00:00    15:49   
1216  160922129    Homeless Complaint  2016-04-01T00:00:00    14:59   
128   160920064  Threats / Harassment  2016-04-01T00:00:00     0:26   
3174  160931527           Petty Theft  2016-04-02T00:00:00    12:07   
8775  160953641    Intoxicated Person  2016-04-04T00:00:00    20:45   
3312  160931756                   470  2016-04-02T00:00:00    13:08   
134   160920070        Noise Nuisance  2016-04-01T00:00:00     0:30   

             CallDateTime Disposition                   Address  \
8289  2016-04-04T16:47:00         CIT       Loomis St/flower St   
404   2016-04-01T06:58:00         GOA   600 Block Of Noriega St   
7000  2016-04-04T05:58:00         HAN       Florida St/treat Av   
3617  2016-04-02T15:49:00         ADV       Shotwell St/18th St   
1216  2016-04-01T14:59:00         ADV     4th St/stillman 2 Suf   
128   2016-04-01T00:26:00          ND      200 Block Of Eddy St   
3174  2016-04-02T12:07:00         REP    0 Block Of Townsend St   
8775  2016-04-04T20:45:00          ND       Dore St/harrison St   
3312  2016-04-02T13:08:00         HAN  2300 Block Of Mission St   
134   2016-04-01T00:30:00         GOA     4100 Block Of 23rd St   

               City State AgencyId  Range      AddressType  
8289  San Francisco    CA        1    NaN     Intersection  
404   San Francisco    CA        1    NaN  Premise Address  
7000  San Francisco    CA        1    NaN     Intersection  
3617  San Francisco    CA        1    NaN     Intersection  
1216            NaN    CA        1    NaN     Geo-Override  
128   San Francisco    CA        1    NaN  Premise Address  
3174  San Francisco    CA        1    NaN  Premise Address  
8775  San Francisco    CA        1    NaN     Intersection  
3312  San Francisco    CA        1    NaN  Premise Address  
134   San Francisco    CA        1    NaN  Premise Address

In [3]:
'''Observar el tipo de variables que contiene el dataframe'''
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10051 entries, 0 to 10050
Data columns (total 12 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   CrimeId                10051 non-null  int64  
 1   OriginalCrimeTypeName  10051 non-null  object 
 2   OffenseDate            10051 non-null  object 
 3   CallTime               10051 non-null  object 
 4   CallDateTime           10051 non-null  object 
 5   Disposition            10051 non-null  object 
 6   Address                10051 non-null  object 
 7   City                   9730 non-null   object 
 8   State                  10048 non-null  object 
 9   AgencyId               10051 non-null  object 
 10  Range                  0 non-null      float64
 11  AddressType            10051 non-null  object 
dtypes: float64(1), int64(1), object(10)
memory usage: 942.4+ KB


In [4]:
'''Detectar valores nulos y su conteo'''
df.isnull().sum()

CrimeId                      0
OriginalCrimeTypeName        0
OffenseDate                  0
CallTime                     0
CallDateTime                 0
Disposition                  0
Address                      0
City                       321
State                        3
AgencyId                     0
Range                    10051
AddressType                  0
dtype: int64

In [5]:
'''Observar la cantidad de filas en el dataset.'''
df.shape

(10051, 12)

# Primer error (Eliminar columnas que no aportan informacion)

In [6]:
'''Eliminar la columna "Range" ya que observando los registros se observa que no posee ningun dato.'''
df = df.drop(['Range'], axis=1)
df.sample(10)

CrimeId OriginalCrimeTypeName          OffenseDate CallTime  \
9799  160963100                   601  2016-04-05T00:00:00    17:52   
453   160920619                Camper  2016-04-01T00:00:00     7:31   
4360  160933722          Traffic Stop  2016-04-02T00:00:00    22:29   
2509  160930314              Physical  2016-04-02T00:00:00     2:11   
9388  160961405          Traffic Stop  2016-04-05T00:00:00    11:18   
4851  160940491         Audible Alarm  2016-04-03T00:00:00     3:07   
846   160921329    Intoxicated Person  2016-04-01T00:00:00    11:15   
6210  160942876          Passing Call  2016-04-03T00:00:00    18:10   
7009  160950441          Passing Call  2016-04-04T00:00:00     6:19   
9870  160963428         Audible Alarm  2016-04-05T00:00:00    19:08   

             CallDateTime   Disposition                         Address  \
9799  2016-04-05T17:52:00           ADV         1100 Block Of Market St   
453   2016-04-01T07:31:00           GOA          700 Block Of Natoma St   
4360  2016-04-02T22:29:00           CIT               Ortega St/19th Av   
2509  2016-04-02T02:11:00           REP          Pine St/leavenworth St   
9388  2016-04-05T11:18:00           CIT            23rd St/hampshire St   
4851  2016-04-03T03:07:00           HAN          900 Block Of Kearny St   
846   2016-04-01T11:15:00           HAN         1900 Block Of Irving St   
6210  2016-04-03T18:10:00  Not recorded         1100 Block Of Market St   
7009  2016-04-04T06:19:00           HAN  2400-3100 Blk Of San Bruno Ave   
9870  2016-04-05T19:08:00           CAN             0 Block Of Grace St   

               City State AgencyId      AddressType  
9799  San Francisco    CA        1  Premise Address  
453   San Francisco    CA        1  Premise Address  
4360  San Francisco    CA        1     Intersection  
2509  San Francisco    CA        1     Intersection  
9388  San Francisco    CA        1     Intersection  
4851  San Francisco    CA        1  Premise Address  
846   San Francisco    CA        1  Premise Address  
6210  San Francisco    CA        1  Common Location  
7009            NaN    CA        1     Geo-Override  
9870  San Francisco    CA        1  Premise Address

In [21]:
'''Eliminar la columna "CallDateTime" ya que observando 
los registros es la misma informacion ya presentada en 
otras columas.'''
df = df.drop(['CallDateTime'], axis=1)
df.sample(10)

CrimeId OriginalCrimeTypeName          OffenseDate CallTime  \
1865  160923298                   213  2016-04-01T00:00:00    19:47   
8531  160953226                   811  2016-04-04T00:00:00    18:36   
5164  160941065          Passing Call  2016-04-03T00:00:00     9:17   
6347  160943099           Pay Dispute  2016-04-03T00:00:00    19:26   
8478  160953125          Traffic Stop  2016-04-04T00:00:00    18:12   
7682  160951646          Traffic Stop  2016-04-04T00:00:00    12:17   
3391  160931893          Passing Call  2016-04-02T00:00:00    13:48   
7078  160950557        Noise Nuisance  2016-04-04T00:00:00     7:15   
8432  160953049       Fight No Weapon  2016-04-04T00:00:00    17:57   
5750  160942075          Traffic Stop  2016-04-03T00:00:00    14:20   

     Disposition                         Address           City State  \
1865         UTL             Lombard St/jones St  San Francisco    CA   
8531         HAN        2500 Block Of Steiner St  San Francisco    CA   
5164         ADV        600 Block Of Columbus Av  San Francisco    CA   
6347         ADV      Fillmore St/golden Gate Av  San Francisco    CA   
8478         CIT        Barneveld Av/mckinnon Av  San Francisco    CA   
7682         CIT               Brannan St/8th St  San Francisco    CA   
3391         HAN  300 Block Of John F Shelley Dr  San Francisco    CA   
7078         GOA         300 Block Of Battery St  San Francisco    CA   
8432         REP        2100 Block Of Oakdale Av  San Francisco    CA   
5750         CIT          200 Block Of London St  San Francisco    CA   

     AgencyId      AddressType  
1865        1     Intersection  
8531        1  Premise Address  
5164        1  Common Location  
6347        1     Intersection  
8478        1     Intersection  
7682        1     Intersection  
3391        1  Common Location  
7078        1  Premise Address  
8432        1  Premise Address  
5750        1  Premise Address

# Segundo error (IDs duplicados)

In [7]:
'''Conteo de ids duplicados, y verificacion de los mismos.'''
print("CrimeId: ",df['CrimeId'].duplicated().sum())
df[df.duplicated(subset=['CrimeId'],keep=False)];

CrimeId:  4


Como se aprecia, existen 4 registros repetidos los cuales analizamos y se tratan de 2 IDs que se registran en 3 ocasiones cada uno, de modo que el sistema asigna 4 repeticiones. Vamos a corregir este error.

In [8]:
'''Realizamos una función la cual crea un set de la lista
de los datos duplicados existentes en la columna CrimeId'''
duplic = df.index[df.duplicated(subset=['CrimeId'],keep=False)].tolist()
maximo = df['CrimeId'].max()

In [9]:
'''Mediante el uso de un ciclo for el cual debe 
recorrer la lista de duplicados. A cada duplicado
se le asigna un nuevo Id, el cual es el valor 
maximo de los IDs aumentado en 1.'''
for r in duplic:
  maximo += 1
  df['CrimeId'].iloc[r] = maximo

C:\Users\User\AppData\Local\Temp\ipykernel_16572\2075909998.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['CrimeId'].iloc[r] = maximo


In [10]:
'''Verificamos si se cambiaron los datos'''
dup = df[df['CrimeId'].duplicated(keep=False)]['CrimeId'].tolist()
dup

[]

# Tercer error (Rellenar ciudades por valores)

In [11]:
'''Se reemplaza los valores vacios que existen 
en las columnas City por valores en Cero, 
ya que al tener valores nulos pueden generar
errores adicionales en este contexto.'''

df[['City', 'State']] = df[['City','State']].fillna(value=0)
df

CrimeId OriginalCrimeTypeName          OffenseDate CallTime  \
0      160903280     Assault / Battery  2016-03-30T00:00:00    18:42   
1      160912272    Homeless Complaint  2016-03-31T00:00:00    15:31   
2      160912590             Susp Info  2016-03-31T00:00:00    16:49   
3      160912801                Report  2016-03-31T00:00:00    17:38   
4      160912811                   594  2016-03-31T00:00:00    17:42   
...          ...                   ...                  ...      ...   
10046  160964210      Well Being Check  2016-04-05T00:00:00    23:38   
10047  160964216          Traffic Stop  2016-04-05T00:00:00    23:43   
10048  160964227          Traffic Stop  2016-04-05T00:00:00    23:48   
10049  160964229          Passing Call  2016-04-05T00:00:00    23:48   
10050  160964249          Traffic Stop  2016-04-05T00:00:00    23:54   

              CallDateTime   Disposition                  Address  \
0      2016-03-30T18:42:00           REP  100 Block Of Chilton Av   
1      2016-03-31T15:31:00           GOA  2300 Block Of Market St   
2      2016-03-31T16:49:00           GOA  2300 Block Of Market St   
3      2016-03-31T17:38:00           GOA      500 Block Of 7th St   
4      2016-03-31T17:42:00           REP       Beale St/bryant St   
...                    ...           ...                      ...   
10046  2016-04-05T23:38:00           GOA    Hollister Av/hawes St   
10047  2016-04-05T23:43:00           ADV   California St/mason St   
10048  2016-04-05T23:48:00           CIT   Rosella Ct/onondaga Av   
10049  2016-04-05T23:48:00  Not recorded          0 Blk Herbst Rd   
10050  2016-04-05T23:54:00           ADV        Grant Av/geary St   

                City State AgencyId      AddressType  
0      San Francisco    CA        1  Premise Address  
1      San Francisco    CA        1  Premise Address  
2      San Francisco    CA        1  Premise Address  
3      San Francisco    CA        1  Premise Address  
4      San Francisco    CA        1     Intersection  
...              ...   ...      ...              ...  
10046  San Francisco    CA        1     Intersection  
10047  San Francisco    CA        1     Intersection  
10048  San Francisco    CA        1     Intersection  
10049  San Francisco    CA        1     Geo-Override  
10050  San Francisco    CA        1     Intersection  

[10051 rows x 11 columns]

In [12]:
'''Verificar la redundancia de datos en la columna de City'''
df.groupby('City').count()

CrimeId  OriginalCrimeTypeName  OffenseDate  CallTime  \
City                                                                   
0                  321                    321          321       321   
 S                   1                      1            1         1   
Brisbane             1                      1            1         1   
Daly City            5                      5            5         5   
Presidio             3                      3            3         3   
SAN FRANCISCO        1                      1            1         1   
San Francisco     9665                   9665         9665      9665   
Treasure Isla       51                     51           51        51   
Yerba Buena          3                      3            3         3   

               CallDateTime  Disposition  Address  State  AgencyId  \
City                                                                 
0                       321          321      321    321       321   
 S                        1            1        1      1         1   
Brisbane                  1            1        1      1         1   
Daly City                 5            5        5      5         5   
Presidio                  3            3        3      3         3   
SAN FRANCISCO             1            1        1      1         1   
San Francisco          9665         9665     9665   9665      9665   
Treasure Isla            51           51       51     51        51   
Yerba Buena               3            3        3      3         3   

               AddressType  
City                        
0                      321  
 S                       1  
Brisbane                 1  
Daly City                5  
Presidio                 3  
SAN FRANCISCO            1  
San Francisco         9665  
Treasure Isla           51  
Yerba Buena              3

<h1>Analisis</h1>
Observamos que existian 321 valores nulos que fueron reemplazados por 0 Analizando las direccciones de estas filas, nos podemos dar cuenta que todas pertenecen a la ciudad de San Francisco, asi que reemplazaremos estos Valores por San Francisco. Del mismo modo se observan dos errores, existe una columna con la letra S, probablemente de San Francisco, comprobaremos la direccion de esta para estar seguros de que la ciudad es San Francisco. Asi mismo observamos que existe un dato con mayusculas. Vamos a corregir ese error.

In [13]:
'''Se reemplaza el dato en mayusculas, por minusculas.'''
df['City'] = df['City'].replace('SAN FRANCISCO','San Francisco');

In [14]:
'''Se reemplaza ahora la letra S por la ciudad de San Francisco'''
df['City']= df['City'].replace(' S','San Francisco')

In [15]:
'''Se reemplaza ahora las filas de 0 por la ciudad de San Francisco'''
df['City']= df['City'].replace(0,'San Francisco')

In [16]:
'''Verificar la redundancia de datos en la columna de City otra vez'''
df.groupby('City').count()

CrimeId  OriginalCrimeTypeName  OffenseDate  CallTime  \
City                                                                   
Brisbane             1                      1            1         1   
Daly City            5                      5            5         5   
Presidio             3                      3            3         3   
San Francisco     9988                   9988         9988      9988   
Treasure Isla       51                     51           51        51   
Yerba Buena          3                      3            3         3   

               CallDateTime  Disposition  Address  State  AgencyId  \
City                                                                 
Brisbane                  1            1        1      1         1   
Daly City                 5            5        5      5         5   
Presidio                  3            3        3      3         3   
San Francisco          9988         9988     9988   9988      9988   
Treasure Isla            51           51       51     51        51   
Yerba Buena               3            3        3      3         3   

               AddressType  
City                        
Brisbane                 1  
Daly City                5  
Presidio                 3  
San Francisco         9988  
Treasure Isla           51  
Yerba Buena              3

In [17]:
'''Verificamos nuevamente los valores nulos, 
como se observa existen 3 estados que son nulos.'''
df.isnull().sum()

CrimeId                  0
OriginalCrimeTypeName    0
OffenseDate              0
CallTime                 0
CallDateTime             0
Disposition              0
Address                  0
City                     0
State                    0
AgencyId                 0
AddressType              0
dtype: int64

# Cuarto error (Valores de columna State)

In [18]:
'''Se verifica la columna state para analizar los valores nulos.
Como se observa existen 3 valores que no tienen valor de estado,
asi que corregiremos este error reemplazando este valor por el estado
correspondiente que es CA.'''
df.groupby('State').count()

CrimeId  OriginalCrimeTypeName  OffenseDate  CallTime  CallDateTime  \
State                                                                        
0            3                      3            3         3             3   
CA       10048                  10048        10048     10048         10048   

       Disposition  Address   City  AgencyId  AddressType  
State                                                      
0                3        3      3         3            3  
CA           10048    10048  10048     10048        10048

In [19]:
'''Reemplazo de los valores 0 de la columna estado.'''
df['State']= df['State'].replace(0,'CA')

In [20]:
'''Verificar nuevamente'''
df.groupby('State').count()

CrimeId  OriginalCrimeTypeName  OffenseDate  CallTime  CallDateTime  \
State                                                                        
CA       10051                  10051        10051     10051         10051   

       Disposition  Address   City  AgencyId  AddressType  
State                                                      
CA           10051    10051  10051     10051        10051